keeper/components/dispatcher_state.go

```diff
@@ -0,0 +1,346 @@
+}
+
+func (k *DispatcherComponent) IterateDispatchedAmountsByProtocolID(
+	ctx context.Context,
+	protocolId types.ProtocolID,
+	callback func(string, types.ChainAmountDispatched) bool,
+) {
+	prefix := collections.NewPrefixedQuadRange[uint32, string, string, string](
+		protocolId.Uint32(),
+	)
+
+	err := k.DispatchedAmounts.Walk(
+		ctx,
+		prefix,
+		func(key DispatchedAmountsKey, value types.AmountDispatched) (stop bool, err error) {
+			orbitID := types.OrbitID{}
+			if err := orbitID.FromString(key.K3()); err != nil {
+				return true, err
+			}
+			dispatchedInfo := types.NewChainAmountDispatched(orbitID, value)
+			return callback(key.K2(), *dispatchedInfo), nil
+		},
+	)
+	if err != nil {
+		k.logger.Error("error in IterateDispatchedByProtocolID walking Dispatched")
+	}
+}
+
+func (k *DispatcherComponent) GetDispatchedAmountsByDestinationProtocolID(
+	ctx context.Context,
+	protocolId types.ProtocolID,
+) types.TotalDispatched {
+	totalDispatched := types.TotalDispatched{
+		ChainAmount: make(map[string]types.ChainAmountDispatched),
+	}
+
+	callback := func(sourceCounterpartyId string, dispatchedInfo types.ChainAmountDispatched) bool {
+		totalDispatched.ChainAmount[sourceCounterpartyId] = dispatchedInfo
+		return false
+	}
+
+	k.IterateDispatchedAmountsByDestinationProtocolID(
+		ctx,
+		protocolId,
+		callback,
+	)
+	return totalDispatched
+}
+
+func (k *DispatcherComponent) IterateDispatchedAmountsByDestinationProtocolID(
+	ctx context.Context,
+	protocolId types.ProtocolID,
+	callback func(string, types.ChainAmountDispatched) bool,
+) {
+	rng := collections.NewPrefixedPairRange[uint32, DispatchedAmountsKey](protocolId.Uint32())
+
+	err := k.DispatchedAmounts.Indexes.ByDestinationProtocolID.Walk(
+		ctx,
+		rng,
+		func(
+			indexingKey uint32,
+			indexedKey DispatchedAmountsKey,
+		) (stop bool, err error) {
+			// Get the actual value from the main collection using the indexed key
+			value, err := k.DispatchedAmounts.Get(ctx, indexedKey)
+			if err != nil {
+				return true, err
+			}
+
+			orbitID := types.OrbitID{}
+			if err := orbitID.FromString(indexedKey.K3()); err != nil {
+				return true, err
+			}
+			dispatchedInfo := types.NewChainAmountDispatched(orbitID, value)
+			return callback(indexedKey.K2(), *dispatchedInfo), nil
+		},
+	)
+	if err != nil {
+		k.logger.Error(
+			"error in IterateDispatchedByDestinationProtocolID walking ByDestinationProtocolID index",
+		)
+	}
+}
+
+// ====================================================================================================
+// DispatchCounts
+// ====================================================================================================
+
+func (k *DispatcherComponent) GetDispatchedCounts(
+	ctx context.Context,
+	sourceInfo types.OrbitID,
+	destinationInfo types.OrbitID,
+) uint32 {
+	destinationId := destinationInfo.ID()
+
+	key := collections.Join3(
+		sourceInfo.ProtocolID.Uint32(),
+		sourceInfo.CounterpartyID,
+		destinationId,
+	)
```

Comment:
nit but doesn't really add anything to store as a variable here when the other info is also accessed in the function call


```suggestion
	key := collections.Join3(
		sourceInfo.ProtocolID.Uint32(),
		sourceInfo.CounterpartyID,
		destinationInfo.ID(),
	)
```

Reply:
yeah these are the dirty tracks left from a refactor haha